src/main/java/org/mifos/pheevouchermanagementsystem/service/RedeemVoucherService.java

```diff
@@ -0,0 +1,88 @@
+package org.mifos.pheevouchermanagementsystem.service;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.mifos.pheevouchermanagementsystem.data.RedeemVoucherRequestDTO;
+import org.mifos.pheevouchermanagementsystem.data.RedeemVoucherResponseDTO;
+import org.mifos.pheevouchermanagementsystem.domain.Voucher;
+import org.mifos.pheevouchermanagementsystem.exception.VoucherNotFoundException;
+import org.mifos.pheevouchermanagementsystem.repository.VoucherRepository;
+import org.mifos.pheevouchermanagementsystem.zeebe.ZeebeProcessStarter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.mifos.pheevouchermanagementsystem.util.EncryptVoucher.hashVoucherNumber;
+import static org.mifos.pheevouchermanagementsystem.util.RedemptionStatusEnum.FAILURE;
+import static org.mifos.pheevouchermanagementsystem.util.RedemptionStatusEnum.SUCCESS;
+import static org.mifos.pheevouchermanagementsystem.util.UniqueIDGenerator.generateUniqueNumber;
+import static org.mifos.pheevouchermanagementsystem.util.VoucherStatusEnum.ACTIVE;
+import static org.mifos.pheevouchermanagementsystem.util.VoucherStatusEnum.UTILIZED;
+import static org.mifos.pheevouchermanagementsystem.zeebe.ZeebeVariables.TENANT_ID;
+
+@Service
+public class RedeemVoucherService {
+    private final VoucherRepository voucherRepository;
+    private final ObjectMapper objectMapper;
+    private static final Logger logger = LoggerFactory.getLogger(RedeemVoucherService.class);
+    @Autowired
+    ZeebeProcessStarter zeebeProcessStarter;
+
+    @Autowired
+    public RedeemVoucherService(VoucherRepository voucherRepository, ObjectMapper objectMapper){
+        this.voucherRepository = voucherRepository;
+        this.objectMapper = objectMapper;
+    }
+
+    public RedeemVoucherResponseDTO redeemVoucher(RedeemVoucherRequestDTO redeemVoucherRequestDTO, String registeringInstitutionId){
+        String transactionId =  generateUniqueNumber(10);
+        Voucher voucher = null;
+        try {
+            voucher = voucherRepository.findBySerialNo(redeemVoucherRequestDTO.getVoucherSerialNumber()).orElseThrow(() -> VoucherNotFoundException.voucherNotFound(redeemVoucherRequestDTO.getVoucherSerialNumber()));
```

Comment:
can you move these checks into a separate method (validate)

Reply:
we are fetching the voucher here by serial number